rust/kernel/types.rs

```diff
@@ -91,3 +91,33 @@ impl<T: PointerWrapper + Deref> PointerWrapper for Pin<T> {
         Pin::new_unchecked(T::from_pointer(p))
     }
 }
+
+/// Runs a cleanup function/closure when dropped.
+///
+/// The [`ScopeGuard::dismiss`] function prevents the cleanup function from running.
+pub struct ScopeGuard<T: FnOnce()> {
+    cleanup_func: Option<T>,
```

Comment:
Not even a nit, probably closer to a code style question. Why name the field if it's the only one in the struct? Presumably you could also do:
```rust
pub struct ScopeGuard<T: FnOnce()>(Option<T>);
```
Is this because an explicit name provides more information than `.0` and makes the code more readable?

Reply:
Yeah, we should decide which style to use.

One tiny advantage of non-tuple ones is avoiding tuple constructor shenanigans:
```
struct T(bool);

pub fn f() -> *const i32 {
    T as _
}
```

Reply:
Sometimes this is an advantage, e.g. you can use the tuple struct name in functions such as `map`.

Obviously for non-public structs there's no difference though.

Reply:
I think a simple rule would be to avoid tuple ones unless the tuple constructor is intended to be used, e.g. for newtypes.

Reply:
@ojeda I don't understand yet how your code example may fail. Or which shenanigans you mean?

But I also don't want to hijack @wedsonaf 's PR :)

Reply:
Or maybe if you couldn't think of a good name :)

BTW `cleanup_func` should probably be named like `drop_fn` or `defer_fn`, as it isn't necessarily for cleanups (sometimes you want the code to run regardless of whether Ok or Err is returned, e.g. logging, not necessarily cleanup).

Reply:
Wait, you edited the code example, lol... maybe it makes more sense to me now :)

Reply:
> @ojeda I don't understand yet which shenanigans you mean. Maybe you could clarify?

Tuple structs (and enum variants) names are themselves a function. E.g. `struct Foo(i32, i32)` means that `Foo` can be used as a function `fn(i32, i32) -> Foo`. (unit structs names are themselves a const)

Reply:
> Sometimes this is an advantage, e.g. you can use the tuple struct name in functions such as `map`.

+1 The lint idea is was about explicit casts into a non-function pointers or integer and things like that. We had a case in e.g. https://github.com/Rust-for-Linux/linux/pull/228.

Reply:
How can a type `T` be cast to an `u8`? I tried this in the playground and of course it runs. `f()` seems to return `192_u8`.

Reply:
It's a pointer to integer cast

Reply:
> +1 The lint idea is more about explicit casts into non-function pointers, integer types and things like that. We had a case in #228 (in particular [174610c](https://github.com/Rust-for-Linux/linux/commit/174610c93e460dae02738d966422370a5227cfaf)).

Clippy does have this lint though? https://rust-lang.github.io/rust-clippy/master/#fn_to_numeric_cast_with_truncation

Reply:
> Wait, you edited the code example, lol... maybe it makes more sense to me now :)

Yeah, sorry, I wanted to make it even more shocking :package: 

Before the edit, there was a cast from function pointer to pointer which sometimes could make sense (e.g. to void pointer to come back later). But casting to integer, and small ones at that, is really something that is almost always not intended.

Reply:
i.e. `T` is a type, but it is also the tuple constructor, which is a function, and so it has function pointer type, and thus can be casted as usual. Normal functions have the problem, but it is less stricking:

```
pub fn f() -> u8 {
    42
}

pub fn g() -> u8 {
    f as _
}
```

So there are things that could be diagnostics or lints: 1. casting function pointers to integers (small ones in particular), I found https://github.com/rust-lang/rust/issues/42915 for something similar; and 2. casting tuple constructors in general.

Reply:
Edited a couple times.

Reply:
GitHub shenanigans here. While I was typing my messages, GitHub was not refreshing, so I could not see your messages which interleaved with mine.

Reply:
That is really shocking! This would be a great interview question to weed out people like me :)

I still don't understand where the `192_u8` comes from. See [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f17019be791908431b0ec5f621186b6f).

Reply:
> Clippy does have this lint though? https://rust-lang.github.io/rust-clippy/master/#fn_to_numeric_cast_with_truncation

Yeah, that is great. However, it does not seem to warn on the to-pointer case:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=96775946c93138afa9fe13a9f03c629b

I think linting about function pointers to "non-void" pointer casts would be valuable (and it would cover the tuple constructor case). After all, you are basically asking to throw away all type information. If a cast like that is really needed, one can allow the lint in that scope.

Reply:
> GitHub shenanigans here. While I was typing my messages, GitHub was not refreshing, so I could not see your messages which interleaved with mine.

Same has happened to me a few times in the last few days. I am not sure if it is a new update, but I have sent messages without realizing there were new comments.

Reply:
> That is really shocking! This would be a great interview question to weed out people like me :)
> 
> I still don't understand where the `192_u8` comes from. See [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f17019be791908431b0ec5f621186b6f).

It is just the bottom bits of the address: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a36bc1e4b100255e9f2c4e5b7786d5d1

Reply:
So you're saying that Rust leaks memory addresses just like that...? I'm not sure what to think about that.

Reply:
From my point of view, this discussion shouldn't derail @wedsonaf 's PR.
LGTM.

Reply:
> So you're saying that Rust leaks memory addresses just like that...? I'm not sure what to think about that.

Well, it is memory-safe to do so on its own ;)

(More seriously: you are right, in the kernel we definitely want to avoid leaking addresses -- but that is a bit orthogonal to this).

> From my point of view, this discussion shouldn't derail @wedsonaf 's PR. LGTM.

No worries!

(I will be logging off for the weekend myself in a few minutes anyway -- I guess Wedson will add the updated docs on Monday).

Reply:
Have a great week-end. Thanks for updating my understanding of Rust!